app/src/main/java/io/gnosis/safe/ui/transactions/details/view/TxConfirmationsView.kt

```diff
@@ -0,0 +1,203 @@
+
+            when (status) {
+                TransactionStatus.AWAITING_EXECUTION -> {
+                    addExecutionStep(TxExecutionStep.Type.EXECUTE_READY)
+                }
+                TransactionStatus.SUCCESS -> {
+                    if(executor != null) {
+                        addExecutionStep(TxExecutionStep.Type.EXECUTE_DONE)
+                        addAddressItem(executor)
+                    } else {
+                        // fail silently
+                        addExecutionStep(TxExecutionStep.Type.EXECUTE_READY)
+                    }
+                }
+                TransactionStatus.CANCELLED -> {
+                    addExecutionStep(TxExecutionStep.Type.CANCELED)
+                }
+                TransactionStatus.FAILED -> {
+                    addExecutionStep(TxExecutionStep.Type.FAILED)
+                }
+            }
+        }
+    }
+
+    private fun addExecutionStep(type: TxExecutionStep.Type, missingConfirmations: Int? = null) {
+        addView(TxExecutionStep(context).apply {
+            val lp = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
+            lp.setMargins(0, 0, 0, dpToPx(MARGIN_VERTICAL))
+            layoutParams = lp
+            setStep(type, missingConfirmations)
+        })
+    }
+    
+    private fun addAddressItem(address: Solidity.Address) {
+        val addressItem = AddressItem(context)
+        val layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, dpToPx(ADDRESS_ITEM_HEIGHT))
+        layoutParams.setMargins(dpToPx(ADDRESS_ITEM_MARGIN_LEFT), 0, 0, dpToPx(MARGIN_VERTICAL))
+        addressItem.layoutParams = layoutParams
+        addressItem.address = address
+        addView(addressItem)
+    }
+
+    fun clear() {
+        removeAllViews()
+    }
+
+    override fun dispatchDraw(canvas: Canvas) {
+        super.dispatchDraw(canvas)
+
+        var x1 = 0f
+        var y1 = 0f
+        for (i in 0 until childCount - 1) {
+
+            val child1 = getChildAt(i)
+            if (child1 is TxExecutionStep) {
+                x1 = child1.stepIconBottom.x
+                y1 = child1.stepIconBottom.y
+            }
+
+            val child2 = getChildAt(i + 1)
+            if (child2 is TxExecutionStep) {
+                canvas.drawLine(
+                    x1,
+                    y1,
+                    child2.stepIconTop.x,
+                    child2.stepIconTop.y,
+                    linePaint
+                )
+            }
+        }
+    }
+
+    private class TxExecutionStep @JvmOverloads constructor(
+        context: Context,
+        attrs: AttributeSet? = null,
+        defStyleAttr: Int = 0
+    ) : ConstraintLayout(context, attrs, defStyleAttr) {
+
+        var type: Type = Type.CREATED
+            private set
+
+        enum class Type {
+            CREATED,
+            CONFIRMED,
+            CANCELED,
+            FAILED,
+            EXECUTE_WAITING,
+            EXECUTE_READY,
+            EXECUTE_DONE
```

Comment:
`EXECUTE_WAITING`, `EXECUTE_READY` and `EXECUTE_DONE` don't sound very gramatical. It is not important, but also, don't we need to distinguish between awaiting execution and awaiting confirmations?

Reply:
we do - EXECUTE_WAITING represents awaiting confirmations, EXECUTE_READY represents awaiting execution. Those names represent ui better. In each case we have execute-label

Reply:
it's not about grammatical sounding. They have same prefix and are used in place of the same execution step element.

Reply:
Why not use the TransactionStatus enum directly then?

Reply:
it does not map 1:1. You don't have created status. Or confirmed. This is very ui specific